ipi/utils/instools.py

```diff
@@ -3,40 +3,80 @@
 from ipi.utils.messages import verbosity, info
 from ipi.utils import units
 import ipi.utils.mathtools as mt
-import os.path, glob
+import os.path, glob,copy
 
-
-def banded_hessian(h, im, shift=0.001):
+def banded_hessian(h, im, masses=True,shift=0.001):
     """Given Hessian in the reduced format (h), construct
-    the upper band hessian including the RP terms"""
+    the upper band hessian including the RP terms.
+    If masses is True returns hessian otherwise it returns dynmat"""
     nbeads = im.dbeads.nbeads
     natoms = im.dbeads.natoms
+    coef   = im.coef  #new_disc
+
     ii = natoms * 3 * nbeads
     ndiag = natoms * 3 + 1  # only upper diagonal form
 
-    # np.set_printoptions(precision=6, suppress=True, threshold=np.nan, linewidth=1000)
-
-    hnew = np.zeros((ndiag, ii))
+    href = np.zeros((ndiag, ii))
 
     # add physical part
     for i in range(nbeads):
         h_aux = h[:, i * natoms * 3:(i + 1) * natoms * 3]  # Peaks one physical hessian
         for j in range(1, ndiag):
-            hnew[j, (ndiag - 1 - j) + i * natoms * 3:(i + 1) * natoms * 3] = np.diag(h_aux, ndiag - 1 - j)
+            href[j, (ndiag - 1 - j) + i * natoms * 3:(i + 1) * natoms * 3] = np.diag(h_aux, ndiag - 1 - j)
 
     # add spring parts
-
     if nbeads > 1:
         # Diagonal
-        d_corner = im.dbeads.m3[0] * im.omega2
+        if masses:
+            d_corner = im.dbeads.m3[0] * im.omega2
+        else:
+            d_corner = np.ones(im.dbeads.m3[0].shape) * im.omega2
+
         d_0 = np.array([[d_corner * 2]]).repeat(im.dbeads.nbeads - 2, axis=0).flatten()
         diag_sp = np.concatenate((d_corner, d_0, d_corner))
-        hnew[-1, :] += diag_sp
+        href[-1, :] += diag_sp
 
         # Non-Diagonal
         d_out = - d_corner
         ndiag_sp = np.array([[d_out]]).repeat(im.dbeads.nbeads - 1, axis=0).flatten()
-        hnew[0, :] = np.concatenate((np.zeros(natoms * 3), ndiag_sp))
+        href[0, :] = np.concatenate((np.zeros(natoms * 3), ndiag_sp))
+
+    # Add safety shift value
+    href[-1, :] += shift
```

Comment:
It has a value that works fine for all the applications I have tested.
It is not accessible from the input file to avoid anyone doing anything crazy. That means
that the value is going to be modified only by people that they know what they are doing.

Reply:
I' ve just added to the function definiton.